src/lib-python/ioda_conv_engines.py

```diff
@@ -0,0 +1,114 @@
+        # and fill them with the provided data and metadata
+        for Vname, Vvals in LocVars.items():
+            VarName = "{0:s}/{1:s}".format(_metagroup, Vname)
+            # create variable
+            self.obsspace.create_var(VarName,
+                                     dtype=Vvals.dtype,
+                                     dim_list=['nlocs'],
+                                     )
+            # write the data to the file
+            tmpVar = self.obsspace.Variable(VarName)
+            tmpVar.write_data(Vvals)
+            # add var metadata
+            try:
+                for MetaVar, MetaVal in LocMdata[Vname].items():
+                    tmpVar.write_attr(MetaVar, MetaVal)
+            except KeyError:
+                pass  # no metadata for this variable
+            # add var units if exists
+            try:
+                UnitStr = LocUnits[Vname]
+                tmpVar.write_attr("units", UnitStr)
+            except (KeyError, NameError):
+                # add error message here later, eventually all need units!
+                pass
+
+    def WriteObsVars(self, ObsVars, VarDims, VarMdata, VarUnits):
+        # this method will create variables in the ouput obs group and
+        # fill them with the provided data and metadata
+        for VarKey, Vvals in ObsVars.items():
+            # get variable and group names
+            (Vname, Gname) = VarKey
+            VarName = "{0:s}/{1:s}".format(Gname, Vname)
+            # create variable
+            self.obsspace.create_var(VarName,
+                                     dtype=Vvals.dtype,
+                                     dim_list=VarDims[Vname],
+                                     )
+            # write the data to the file
+            tmpVar = self.obsspace.Variable(VarName)
+            tmpVar.write_data(Vvals)
+            # add var metadata
+            try:
+                for MetaVar, MetaVal in VarMdata[Vname].items():
+                    tmpVar.write_attr(MetaVar, MetaVal)
+            except KeyError:
+                pass  # no metadata for this variable
+            # add var units if exists
+            if Gname not in _no_units:
+                try:
+                    UnitStr = VarUnits[Vname]
+                    tmpVar.write_attr("units", UnitStr)
+                except (KeyError, NameError):
+                    # add error message here later, eventually all need units!
+                    pass
+
+    def WriteGlobalAttrs(self, AttrData):
+        # this method will create global attributes from AttrData dictionary
+        for AttrKey, AttrVal in AttrData.items():
+            self.obsspace.write_attr(AttrKey, AttrVal)
+
+    def BuildIoda(self, ObsVars, VarDims, LocVars,
+                  VarMdata, AttrData, VarUnits={}, TestData=None):
+        self.WriteObsVars(ObsVars, VarDims, VarMdata, VarUnits)
+        self.WriteLocVars(LocVars, VarMdata, VarUnits)
```

Comment:
The old library had separate dictionaries for location indexed metadata, variable indexed metadata and record indexed metadata. I think this notion is outdated so there isn't really a need to keep the location indexed metadata in a separate dictionary (at least based on the location, variable, and record idea).

If that's the only reason for keeping ObsVars and LocVars in separate dictionaries, I think it would be fine to merge those together.

Now that we have dimensioned data and are taking great care to manage the dimensions of data, we can put all metadata into the "MetaData" group and their dimensions denote which kind of meta data they are. Eg, indexed by nchans means it is meta data describing channels (such as wavelength values).

Again I think this is good for now, and we can address this (if anything needs to be done) in subsequent PRs. We probably should get to this before the code sprint however.

Reply:
Yes I think the only real difference is that the LocVars function explicitly assumes "MetaData" but we can just follow what is done for the ObsVars function which is more generic. It is up to you if this is done now or in a new PR, I'm happy to change it in this PR if you want, or we can do it next after this is merged so it is a smaller change.

Reply:
If this is a relatively quick change, I'm inclined to do it now since this is a change in the script that is using the library (and there is only the one script to change at this time). Otherwise, let's merge this PR first. Thanks!

Reply:
Ok, let me try it after lunch and if its easy enough I'll just push a new commit.

Reply:
@srherbener take a look at the newest commit and let me know what you think. Tests pass (on Hera) without changing the reference file.

Reply:
Looks great! Thanks for making this change.